worker/peergrouper/worker.go

```diff
@@ -435,53 +437,86 @@ func machineAddresses(machines map[string]*machineTracker) [][]network.Address {
+
 // getMongoSpace updates info with the space that Mongo servers should exist in.
 func (w *pgWorker) getMongoSpace(addrs [][]network.Address) (network.SpaceName, error) {
-	unset := network.SpaceName("")
-
 	stateInfo, err := w.config.State.ControllerInfo()
 	if err != nil {
-		return unset, errors.Annotate(err, "cannot get state server info")
+		return unsetSpace, errors.Annotate(err, "cannot get state server info")
 	}
 
 	switch stateInfo.MongoSpaceState {
 	case state.MongoSpaceUnknown:
 		if !w.config.SupportsSpaces {
 			err := w.config.State.SetMongoSpaceState(state.MongoSpaceUnsupported)
 			if err != nil {
-				return unset, errors.Annotate(err, "cannot set Mongo space state")
+				return unsetSpace, errors.Annotate(err, "cannot set Mongo space state")
 			}
-			return unset, nil
+			return unsetSpace, nil
 		}
 
-		// We want to find a space that contains all Mongo servers so we can
-		// use it to look up the IP address of each Mongo server to be used
-		// to set up the peer group.
-		spaceStats := generateSpaceStats(addrs)
-		if spaceStats.LargestSpaceContainsAll == false {
-			err := w.config.State.SetMongoSpaceState(state.MongoSpaceInvalid)
-			if err != nil {
-				return unset, errors.Annotate(err, "cannot set Mongo space state")
+		// Check to see if a HA space was set in the controller config.
+		space := unsetSpace
+		space, err = w.tryGetMongoSpaceFromConfig(addrs)
+		if err != nil {
+			return space, errors.Annotate(err, "getting Mongo space from config")
+		}
+
+		// If not, try to determine a space automatically.
+		if space == unsetSpace {
+			if space, err = w.tryFindMongoSpace(addrs); err != nil {
+				if err == errSpaceNotSuperset {
+					logger.Warningf(err.Error())
+					err = nil
+				}
+				return space, errors.Trace(err)
 			}
-			logger.Warningf("couldn't find a space containing all peer group machines")
-			return unset, nil
 		}
 
-		spaceName, err := w.config.State.SetOrGetMongoSpaceName(spaceStats.LargestSpace)
-		if err != nil {
-			return unset, errors.Annotate(err, "error setting/getting Mongo space")
+		if space, err = w.config.State.SetOrGetMongoSpaceName(space); err != nil {
+			err = errors.Annotate(err, "error setting/getting Mongo space")
 		}
-		return spaceName, nil
+		return space, err
 
 	case state.MongoSpaceValid:
 		space, err := w.config.State.Space(stateInfo.MongoSpaceName)
 		if err != nil {
-			return unset, errors.Annotate(err, "looking up space")
+			return unsetSpace, errors.Annotate(err, "looking up space")
 		}
 		return network.SpaceName(space.Name()), nil
 	}
 
-	return unset, nil
+	return unsetSpace, nil
+}
+
+// trySetMongoSpaceFromConfig checks whether a value was supplied for
+// juju-ha-space in the controllers config.
+// If so, an attempt is made to use that value as the Mongo space name.
+func (w *pgWorker) tryGetMongoSpaceFromConfig(addrs [][]network.Address) (network.SpaceName, error) {
```

Comment:
why are you passing in the addresses here? Aren't you just getting the space name?

Reply:
Ah, yes. I added then removed code in that method, but left the argument; will fix.